File: src/so/aporia/module/settings/Setting.java
/*
 * Copyright (c) 2025-2026 Aporia.cc Project
 * Distributed under the Aporia.cc Software License Agreement v1.0
 * See LICENSE and COPYRIGHT files in the project root for full text.
 */

package so.aporia.module.settings;

import java.util.function.Supplier;

/**
 * Base class for all module settings.
 */
public abstract class Setting<T> {
    protected final String name;
    protected final String description;
    protected final Supplier<Boolean> visible;
    protected T value;
    
    protected Setting(String name, String description, T defaultValue, Supplier<Boolean> visible) {
        this.name = name;
        this.description = description;
        this.value = defaultValue;
        this.visible = visible;
    }
    
    public String name() { return name; }
    public String description() { return description; }
    public T get() { return value; }
    public void set(T value) { this.value = value; }
    public boolean isVisible() { return visible == null || visible.get(); }
}

File: src/so/aporia/module/settings/Setting.kt
package so.aporia.module.settings

import so.aporia.utils.user.render.core.AporiaRenderer
import so.aporia.utils.user.render.theme.ThemeManager.Theme
import java.util.function.Supplier
import com.chaos.annotation.ChaosNative
@ChaosNative
abstract class Setting<T>(
    val name: String,
    val description: String,
    value: T,
    private val visible: Supplier<Boolean>? = null,
    var category: String = "",
    var hierarchy: Int = 0
) {
    init {
        val cur = SettingCategoryManager.current()
        if (cur != null && category.isEmpty()) {
            category = cur.first; hierarchy = cur.second
        }
    }

    var value: T = value
        protected set(v) {
            field = v
            so.aporia.utils.files.impl.ConfigFile.markDirty()
        }

    open fun isVisible(): Boolean = visible?.get() ?: true
    fun name(): String = name

    open fun displayHeight(dropOpen: Boolean): Float = 14f

    open fun draw(r: AporiaRenderer, x: Float, y: Float, w: Float, theme: Theme, mouseX: Float, mouseY: Float, isDropOpen: Boolean) {}
}
